proposals/0055-csp.md

```diff
@@ -0,0 +1,251 @@
+
+## Server Islands
+
+Server islands are very dynamic because the contents of their rendered scripts depend on `props` and `slots`. Due to this requirement, the server islands must support head propagation. The compiler will ensure that the correct metadata is emitted.
+
+The hashes **will be generated during the rendering phase**. The performance hit is proportional to the number of server islands used in a page.
+
+## User styles
+
+Compiled styles are calculated during the build, and Astro already possesses all the information on the styles imported and rendered by page. During the build, Astro will calculate the hashes of those assets and store them in the manifest. The rendering engine will just read and render them.
+
+## Rendered `<meta>`
+
+The current proposal targets scripts and styles emitted by Astro, so the directives rendered will handle only `style-src` and `script-src`:
+```html
+<meta 
+  http-equiv="Content-Security-Policy"
+  content="style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2'"
+>
+```
+While browsers are able to [merge directives](https://centralcsp.com/articles/csp-meta-tags) from `Response` headers and `meta` tags, Astro will provide runtime APIs so users can push their custom policies based in their requirements.
+
+### Configuration APIs
+
+New configurations will be available in case directives and/or hashes can be applied to all pages
+
+```js
+export default defineConfig({
+    security: {
+        csp: {
+            // or SHA-384, or SHA-256 (default)
+            algorithm: "SHA-512",
+            // Applied only to `script-src` directive
+            strictDynamic: true,
+            // Additional directives added to the final content
+            directives: [
+                "default-src: 'self'",
+                "img-src 'self'"
+            ],
+            // List of hashes for `style-src`. 
+            styleHashes: [
+                "sha512-hash1",
+                "sha256-hash1",
+                "sha384-hash1",
+            ],
+            // List of hashes for `script-src`. 
+            scriptHashes: [
+                "sha512-hash1",
+                "sha256-hash1",
+                "sha384-hash1",
+            ],
+        }
+    }
+})
+```
+
+### Runtime APIs
+
+The following paragraph will describe possible runtime APIs that users can use to customise their CSP policies. When `Astro.` is used, it's intended that also `APIContext` is covered by the implementation.
+
+#### `Astro.insertCspDirective(directive: string)`
+
+Users can add an entire directive. Astro won't check if the directive is grammatically correct.
+
+```js
+Astro.insertCspPolicy(`image-src '${Astro.site}'`)
+```
+Result
+```html
+<meta 
+  http-equiv="Content-Security-Policy"
+  content="image-src 'https://example.com'; style-src 'self' 'sha256-hash1'; script-src 'self' 'sha256-hash2'"
```

Comment:
Does order matter?

Reply:
Can you be more specific with your query? The order of what :)

Reply:
Sorry! I meant does it matter if it's first image-src, then style-src, then script-src or can it be in any order? eg. image-src at the end

Reply:
It won't matter